librad/src/net/protocol/io/send/request_response.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::net::SocketAddr;
+
+use futures::{SinkExt as _, TryStreamExt as _};
+use futures_codec::Framed;
+
+use crate::net::{
+    codec::CborCodec,
+    connection::{RemoteAddr as _, RemotePeer as _},
+    protocol::{error, interrogation, quic, upgrade},
+};
+
+pub trait Pairing {
+    type Response;
+    type Upgrade: Into<upgrade::UpgradeRequest>;
+    const UPGRADE: Self::Upgrade;
+}
+
+impl Pairing for interrogation::Request {
```

Comment:
Are there other `Pairing`s? I was expecting to possibly see `Git`, `Gossip`, and `Membership` ones based on the different upgrades we've defined

Reply:
Yeah but they’re not allowed in a request/response pattern. I was briefly tempted to generalise the `send_rpc` function in a similar way, but then decided to leave it to a later patch, because it would add quite some diff noise to this one.